### PR TITLE
[RFC] `handle_certificate` should return `SignedTransactionEffects` only

### PR DESCRIPTION
**Pull request by lxfind**
Currently `handle_certificate` returns `TransactionResponseInfo`, which contains 3 optional values: signed transaction, certificate and signed effects. This is bad in several ways:
1. When executing a certificate, we only need the signed effects. Returning all 3 values is wasting network response traffic and requires extra db reads.
2. It potentially creates confusions, for instance, the signed effects in the response can never be None; or what happens if the signed transaction is None?
3. It makes it difficult to extend in the future, when we support finalized transactions.
4. It requires reading from epoch table when making the result.

This PR changes the return type of `handle_certificate` to be `SignedTransactionEffects` only.

### CODE REVIEW

File: crates/sui-core/src/authority.rs [hunk 1]
```diff
@@ -738,15 +733,15 @@ impl AuthorityState {
...
         certificate: &VerifiedCertificate,
-    ) -> SuiResult<VerifiedTransactionInfoResponse> {
+    ) -> SuiResult<SignedTransactionEffects> {
         let digest = *certificate.digest();
         // The cert could have been processed by a concurrent attempt of the same cert, so check if
         // the effects have already been written.
         // If the cert is finalized in a previous epoch, it will be re-signed
         // with current epoch info and returned.
-        if let Some(info) = self.get_tx_info_already_executed(&digest).await? {
+        if let Some(signed_effects) = self.database.get_signed_effects(&digest)? {
```
Comment (longbowlu) [inline crates/sui-core/src/authority.rs:742]: we need to re-sign the effects as the comment says

File: crates/sui-core/src/safe_client.rs [hunk 2]
```diff
@@ -187,6 +187,42 @@ impl<C> SafeClient<C> {
...
+        digest: &TransactionDigest,
+        signed_effects: &SignedTransactionEffects,
+        expected_effects_digest: Option<&TransactionEffectsDigest>,
+    ) -> SuiResult {
+        // Check it has the right signer
+        fp_ensure!(
+            signed_effects.auth_sig().authority == self.address,
+            SuiError::ByzantineAuthoritySuspicion {
+                authority: self.address,
+                reason: "Unexpected validator address in the signed effects signature".to_string()
```
Comment (longbowlu) [inline crates/sui-core/src/safe_client.rs:201]: include the actual address?

File: crates/sui-core/src/unit_tests/authority_tests.rs [hunk 3]
```diff
@@ -1735,11 +1732,11 @@ async fn test_handle_confirmation_transaction_bad_sequence_number() {
 
     // Explanation: providing an old cert that has already been applied
     //              returns a Ok(_) with info about the new object states.
-    let response = authority_state
+    // TODO: Why was signed_effects None?
```
Comment (longbowlu) [inline crates/sui-core/src/unit_tests/authority_tests.rs:1735]: this TODO is not clear, better to add a PR ID for future reference

**Comment by longbowlu** [comment]
after this and https://github.com/MystenLabs/sui/pull/7003/ lands, we can remove `signed_effects` in the `TransactionResponseInfo` as well

File: crates/sui-core/src/authority.rs (hunk 1 above)
Comment (lxfind) [reply crates/sui-core/src/authority.rs:742]: Good catch

File: crates/sui-core/src/unit_tests/authority_tests.rs (hunk 3 above)
Comment (lxfind) [reply crates/sui-core/src/unit_tests/authority_tests.rs:1735]: Turns out this test is disabled

**Comment by lxfind** [comment]
> after this and https://github.com/MystenLabs/sui/pull/7003 lands, we can remove signed_effects in the TransactionResponseInfo as well

Probably not. It can serve as a fast path, in case you are calling handle_transaction on a transaction that has already been executed

**Review by mystenmark — APPROVED** [review]
Great idea, thanks!

**Pull request merged**